### ISSUE TITLE
Add additional permission to GPL in opencl and blas directories

### ISSUE THREAD

**Issue by borg323** [LeelaChessZero/lc0#657]
Following the recent change to Leela zero licensing (https://github.com/gcp/leela-zero/commit/808bb43df34654d357be2dd278eba19c03f07094) we can similarly update the license in the opencl, blas and shared directories that reuse Leela zero code. This will allow the distribution of all backends compiled in the same executable. The additional permission added by Leela zero is:
```
    Additional permission under GNU GPL version 3 section 7

    If you modify this Program, or any covered work, by linking or
    combining it with NVIDIA Corporation's libraries from the
    NVIDIA CUDA Toolkit and/or the NVIDIA CUDA Deep Neural
    Network library and/or the NVIDIA TensorRT inference library
    (or a modified version of those libraries), containing parts covered
    by the terms of the respective license agreement, the licensors of
    this Program grant you additional permission to convey the resulting
    work.
```
This is compatible with the current license but also adds permission to link against TensorRT. To add this we need permission from all contributors in the code in these directories (in alphabetical order):
@borg323 (OK for all contributions)
@dubslow 
@frpays 
@mooskagh 
@ra1u 
@ttl

Please add a comment saying whether you **agree or not with this change**, and whether you **agree to extend this to all your contributions** in case we want to add a TensorRT backend in the future. Also please indicate if I missed any contributors. I will keep this list updated with the latest information.

**Comment by ra1u** [comment]
Ok for all contributions.

**Comment by ra1u** [comment]
Just to be clear, can someone explain in simple terms how  line "the licensors of this Program grant you additional permission to convey the resulting work."  is to be interpreted?

**Comment by borg323** [comment]
I think the explanation in the GPL FAQ is the best we can hope for: <https://www.gnu.org/licenses/gpl-faq.en.html#GPLIncompatibleLibs>

**Comment by frpays** [comment]
Ok for all contributions.

**Comment by mooskagh** [comment]
Ok for all contributions!

**Comment by Naphthalin** [comment]
Any update on this? @dubslow doesn't seem to have responded.

**Comment by borg323** [comment]
This needs a careful code review, what we have so far may be sufficient.

**Comment by nguyenpham** [comment]
I am always a bit scared whenever we do modify default licenses since that doesn’t look as safe and simple as original ones. However, this project is an open source, everything is clear and only for communities. Thus I support things which may help more for communities:
OK for all contributions

**Comment by borg323** [comment]
@nguyenpham in this case we are using the wording suggested by the FSF for this exact scenario, Note also that the additional permission is optional, so everybody is free to use the source code under the standard GPL v3 if they want.

**Comment by borg323** [comment]
Since this was opened, I see also @ddobbelaere and @gsobala have added code in the relevant directories. What do you think about the suggested change?

**Comment by ddobbelaere** [comment]
Ok for all contributions!

**Comment by Naphthalin** [comment]
Was this issue completed? @borg323

**Comment by borg323** [comment]
No, this needs code review to see if dubslow made any code contributions - doesn't seem like it but this is not conclusive. Also need to for later code changes.
